### PR TITLE
DA1469x clock control driver

### PR DESCRIPTION
**Pull request by blauret**
This commit adds basic support for the clock controller used in SmartBond Da1469x MCUs family

Signed-off-by: Ben Lauret <ben.lauret.wm@renesas.com>

### CODE REVIEW

File: dts/bindings/clock/renesas,smartbond-clock.yaml [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+description: Renesas Smartbond clock control node
+
+compatible: "renesas,smartbond-clock"
+
+include: base.yaml
+
+properties:
+  reg:
+    required: false
```
Comment (galak) [inline dts/bindings/clock/renesas,smartbond-clock.yaml:12]: remove `required: false` its the default.

**Review by galak — COMMENTED** [review]
How do you expect the request/release api's to be used?

File: dts/bindings/clock/renesas,smartbond-clock.yaml (hunk 1 above)
Comment (blauret) [reply dts/bindings/clock/renesas,smartbond-clock.yaml]: Done

**Comment by blauret** [comment]
> How do you expect the request/release api's to be used?

There are multiple use cases. The PLL is required to run the USB interface (PR is coming for the USB). It will request the PLL and as long as the USB is enabled it keeps it running. If you disable the USB it allows the system to fall back to the 32MHz. 

The other use case is for application requiring more computing power. The application can request the PLL for faster processing. It's specificaly useful in conjunction with the LCD controller. An example is when you get an interruption to turn on the screen, application requests the PLL for faster update. When the screen turns off it is released to save power.

File: drivers/clock_control/clock_control_smartbond.c [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+LOG_MODULE_REGISTER(clock_control, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
+
+#define DT_DRV_COMPAT smartbond_clock
+
+static atomic_t pll_lock_count;
+
+struct clocks_dev_data {
+	/* RCX rate, can be measured at runtime */
+	uint32_t rcx_rate;
+	/* RCX rate, can be measured at runtime */
```
Comment (ydamigos) [inline drivers/clock_control/clock_control_smartbond.c:23]: RCX -> RC32K

File: dts/arm/renesas/smartbond/da1469x.dtsi [hunk 3]
```diff
@@ -19,6 +20,66 @@
...
+				compatible = "fixed-clock";
+				clock-frequency = <DT_FREQ_M(32)>;
+				#clock-cells = <0>;
+				status = "disabled";
+			};
+			pll: pll {
+				compatible = "fixed-clock";
+				clock-frequency = <DT_FREQ_M(96)>;
+				#clock-cells = <0>;
+				status = "disabled";
```
Comment (ydamigos) [inline dts/arm/renesas/smartbond/da1469x.dtsi:61]: We should state the dependency between PLL and XTAL32M. Right now if xtal32m status is disabled and pll status is ok, we could enable PLL in `smartbond_clocks_init()` and disable XTAL32M.

**Review by ydamigos — CHANGES_REQUESTED** [review]

File: drivers/clock_control/clock_control_smartbond.c [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+
+	switch (*clk) {
+	case DT_DEP_ORD(DT_NODELABEL(rc32k)):
+		*rate = data->rc32k_rate;
+		break;
+	case DT_DEP_ORD(DT_NODELABEL(rcx)):
+		*rate = data->rcx_rate;
+		break;
+	case DT_DEP_ORD(DT_NODELABEL(xtal32k)):
+		*rate = DT_PROP(DT_NODELABEL(xtal32k), clock_frequency);
```
Comment (ydamigos) [inline drivers/clock_control/clock_control_smartbond.c:197]: It would be good to configure XTAL32K before enabling it.

**Comment by MaureenHelm** [comment]
@blauret ping

**Comment by kasjer** [comment]
Clock control code was update notable differences to previous iteration:
- renesas/hal (from separate repository) is used for actual xtal/rc/pll manipulation
- PLL request functions were removed and now PLL is enabled in DT
- additional static assert is added to prevent case when PLL is enabled but XTAL32M is not (noticed by @ydamigos)

**Comment by blauret** [comment]
@MaureenHelm, thanks for checking on this one. We updated the PR and we believe all prior comments have been addressed

**Review by andrzej-kaczmarek — APPROVED** [review]

**Review by sjanc — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**